### PR TITLE
Draft gRPC

### PR DESCRIPTION
**Pull request by fibonacci1729**

### CODE REVIEW

File: cmd/draft/draft.go [hunk 1]
```diff
@@ -108,13 +108,13 @@ func teardown() {
...
-	}
-	client, err := draft.NewFromString(draftHost, nil)
-	if err != nil {
-		panic(err)
+func ensureDraftClient(client *draft.Client) *draft.Client {
+	if client == nil {
+		return draft.NewClient(&draft.ClientConfig{
+			ServerAddr: draftHost,
+			Stdout:     os.Stdout,
+			Stderr:     os.Stderr,
```
Comment (bacongobbler) [inline cmd/draft/draft.go]: we also might want to add a `Stdin` field here as well, as `draft init` and some of the Draft plugins read from stdin.

Comment (fibonacci1729) [inline cmd/draft/draft.go]: True. However, I think we should hold off until (if) we need it. Right now, the only 2 commands that use the draft client are `version` and `up`.

File: cmd/draftd/start.go [hunk 2]
```diff
@@ -71,43 +71,36 @@ func newStartCmd(out io.Writer) *cobra.Command {
...
-	if err != nil {
-		return fmt.Errorf("failed to create server at %s: %v", c.listenAddr, err)
-	}
-	server.DockerClient = dockerClient
-	server.RegistryAuth = c.registryAuth
-	server.RegistryOrg = c.registryOrg
-	server.RegistryURL = c.registryURL
-	server.Basedomain = c.basedomain
-	server.HelmClient = helm.NewClient(helm.Host(c.tillerURI))
-	server.KubeClient = kubeClientset
```
Comment (bacongobbler) [inline cmd/draftd/start.go]: 👍 nice to see this spaghetti code removed.

File: cmd/draft/draft.go (hunk 1 above)
Comment (bacongobbler) [inline cmd/draft/draft.go]: good point. I see what this does now after a more thorough review. I thought this was a replacement for `newRootCmd` in cmd/draft/draft.go, but it looks more like this is the stdout and stderr from the server's output.

File: pkg/draft/server.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	"fmt"
+	"github.com/Azure/draft/pkg/rpc"
+	"golang.org/x/net/context"
+	"net"
+	"sync"
+)
+
+// kubernetes imports
+import (
+	// 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (bacongobbler) [inline pkg/draft/server.go]: note to self that we will need to remove these before we merge.

**Comment by bacongobbler** [comment]
I'm seeing this when deploying draftd via `draft init` with the new client:

```
2017/08/08 18:34:12 transport: http2Server.HandleStreams received bogus greeting from client: "GET /ping HTTP/1.1\r\nHost"
```

This is causing the pod to fail its readiness health check, so it's always in a `1/2 CrashLoopBackOff` state.

EDIT: and now I see the new commit fixing this. Thanks!

**Comment by bacongobbler** [comment]
So I was able to get the stages running locally, but when Helm is attempting to install the package I'm seeing this error pop up:

```
Pushing Docker Image: success
Releasing Application: started
Releasing Application: Release "elevated-skunk" does not exist. Installing it now.
Releasing Application: failure: could not install release: rpc error: code = Unknown desc = release elevated-skunk failed: Ingress.extensions "elevated-skunk-elevated-skunk" is invalid: spec.rules[0].host: Invalid value: "elevated-skunk.86d48673956bc03a6d0d6101ae546b386351cd63%!(EXTRA string=k8s.local)": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

I'll see if I can do a bit more diving before our pairing session later today :)

File: pkg/draft/context.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	imgtag := fmt.Sprintf("%.20x", h.Sum(nil))
+	prefix := s.cfg.Registry.Org
+	if prefix != "" {
+		prefix += "/"
+	}
+	image := fmt.Sprintf("%s/%s%s:%s", s.cfg.Registry.URL, prefix, req.AppName, imgtag)
+
+	// inject certain values into the chart such as the registry location,
+	// the application name, and the application version.
+	tplstr := "image.name=%s,image.org=%s,image.registry=%s,image.tag=$s,basedomain=%s"
```
Comment (bacongobbler) [inline pkg/draft/context.go]: typo: `tag=$s` should be `tag=%s`

Comment (bacongobbler) [inline pkg/draft/context.go]: also missing here is a recently merged PR (sorry!)

#234

File: pkg/draft/context.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	imgtag := fmt.Sprintf("%.20x", h.Sum(nil))
+	prefix := s.cfg.Registry.Org
+	if prefix != "" {
+		prefix += "/"
+	}
+	image := fmt.Sprintf("%s/%s%s:%s", s.cfg.Registry.URL, prefix, req.AppName, imgtag)
+
+	// inject certain values into the chart such as the registry location,
+	// the application name, and the application version.
+	tplstr := "image.name=%s,image.org=%s,image.registry=%s,image.tag=%s,basedomain=%s"
```
Comment (bacongobbler) [inline pkg/draft/context.go]: I completely forgot in the pairing session! Can you please add `ondraft=true` here so we don't accidentally lose #234? Thanks!

Comment (fibonacci1729) [inline pkg/draft/context.go]: 👍

**Comment by bacongobbler** [comment]
Awesome. Looks like just a few linter fixes and we're good to merge.

**Comment by errordeveloper** [comment]
<3

**Pull request merged**